lib/src/Uncertainty/Algorithm/Simulation/CrossEntropyImportanceSampling.cxx

```diff
@@ -0,0 +1,245 @@
+  {
+    ++iterationNumber ;
+    
+    Point currentAuxiliaryDistributionParameters = auxiliaryDistributionParameters;
+
+    // Drawing of samples using auxiliary density and evaluation on limit state function
+    auxiliaryInputSample = Sample(0, initialDistribution_.getDimension());
+    auxiliaryOutputSample = Sample(0, 1);
+
+
+
+    for (UnsignedInteger i = 0; i < getMaximumOuterSampling(); ++i)
+    {
+      const Sample blockSample(auxiliaryDistribution_.getSample(getBlockSize()));
+      auxiliaryInputSample.add(blockSample);
+      auxiliaryOutputSample.add(computeOutputSamples(blockSample));
+
+      if (stopCallback_.first && stopCallback_.first(stopCallback_.second))
+        throw InternalException(HERE) << "User stopped simulation";
+    }
+
+    // Computation of current quantile
+    currentQuantile = auxiliaryOutputSample.computeQuantile(rhoQuantile_)[0];
+    
+    // If failure probability reached, stop the adaptation
+    if (getEvent().getOperator()(currentQuantile, getEvent().getThreshold()))
+    {
+      currentQuantile = getEvent().getThreshold();
+    }
+    else
+    {
+    Indices indiceCritic(0);
+    for (UnsignedInteger i = 0; i < auxiliaryInputSample.getSize(); ++i)
+      {
+        const Bool weightBool = getEvent().getOperator()(auxiliaryOutputSample(i, 0), currentQuantile);
+        if (weightBool)
+          indiceCritic.add(i);
+      } 
+
+   // Extract the relevant sample
+   const Sample auxiliaryCriticInputSamples(auxiliaryInputSample.select(indiceCritic));
+  
+  
+   // Optimize auxiliary distribution parameters
+   Point auxiliaryDistributionParameters = optimizeAuxiliaryDistributionParameters(auxiliaryCriticInputSamples);
+  
+   // Update auxiliary Distribution Parameters
+   updateAuxiliaryDistribution(auxiliaryDistributionParameters); 
+
+    //throw InternalException(HERE) << "User stopped simulation"; 
+      
+    }
+
+    if (stopCallback_.first && stopCallback_.first(stopCallback_.second))
+      throw InternalException(HERE) << "User stopped simulation";
+
+  } // while
+
+  // Find failure sample indices
+  Indices indicesCritic(0);
+  for (UnsignedInteger i = 0; i < auxiliaryOutputSample.getSize(); ++i)
+  {
+    // Find failure Points
+    if (getEvent().getOperator()(auxiliaryOutputSample(i, 0), getEvent().getThreshold()))
+      indicesCritic.add(i);
+  } // for i
+
+  const Sample resp_sampleCritic(auxiliaryOutputSample.select(indicesCritic));
+  const Sample sampleCritic(auxiliaryInputSample.select(indicesCritic));
+
+  // Evaluate initial log PDF in parallel on failure sample
+  Sample logPDFInitCritic(initialDistribution_.computeLogPDF(sampleCritic));
+
+  // Evaluate auxiliary log PDF in parallel on failure sample
+  Sample logPDFAuxiliaryCritic(auxiliaryDistribution_.computeLogPDF(sampleCritic));
+
+  Scalar sumPdfCritic = 0.0;
+  for(UnsignedInteger i = 0; i < indicesCritic.getSize(); ++i)
+  {
+    sumPdfCritic += std::exp(logPDFInitCritic(i, 0) - logPDFAuxiliaryCritic(i, 0));
```

Comment:
This looks suspicious to me because $s  = \exp\left[\log(a) - \log(b)\right] = \exp\left[\log(a/b)\right] = a / b$ if $a > 0$ and $b > 0$.
Can we evaluate the ratio of PDFs directly in this case or do we only have access to the logPDF?

Reply:
This was suggested in the development of NAIS because the magnitude of the pdf can be very small (allows to improve the numerical efficiency).

Reply:
In this particular case, we do have access to the PDF: using the logPDF is not mandatory, but a design choice. 

I have doubts about the accuracy improvement because a catastrophic cancellation can occur with a subtraction, but cannot with a division. This is because the sum operator y = a + b can be ill-conditionned if a and b are of opposite signs and y is close to zero, but the ratio y = a / b is always well conditioned. 

Perhaps a mathematical description would help. According to the code, the variable `sumPdfCritic` contains the sum of ratio of PDF values. I would have less doubts if the variable sumPdfCritic was made of a large sum of log-PDF values, then finally exponentialized. But this is not what is done here: the exponential function is evaluated directly after the difference of logarithms.

Do you have a formal proof or example of the loss of accuracy when each pdfs values are close to zero?

Reply:
Hi,

To develop Cross Entropy, we were based on what was done on NAIS. As this part of the code is the same as NAIS, we just took the same design choices as those that were made in this algorithm. 

Maybe @regislebrun could have some opinion on this matter?